athenzdomains.go

```
@@ -0,0 +1,183 @@
+			}
+			actionRegex, err := regexp.Compile("^" + replacer.Replace(strings.ToLower(assertion.Action)) + "$")
+			if err != nil {
+				fmt.Printf("Error occurred when converting assertion action into regex format. Error: %v", err)
+			}
+			simpleAssert := SimpleAssertion{
+				resource: resourceRegex,
+				action:   actionRegex,
+				effect:   effect,
+			}
+			crMap.RoleToAssertion[assertion.Role] = append(crMap.RoleToAssertion[assertion.Role], simpleAssert)
+		}
+
+	}
+}
+
+func (c *Cache) addObj(domainMap map[string]CRMap, item *v1.AthenzDomain) {
+	c.lock.Lock()
+	domainName := item.ObjectMeta.Name
+	_, ok := domainMap[domainName]
+	if !ok {
+		roleToPrincipals := make(map[string][]*zms.RoleMember)
+		roleToAssertion := make(map[string][]SimpleAssertion)
+		crMap := CRMap{
+			RoleToPrincipals: roleToPrincipals,
+			RoleToAssertion:  roleToAssertion,
+		}
+		domainMap[domainName] = crMap
+	}
+	parseData(domainMap, domainName, item)
+	c.lock.Unlock()
+}
+
+func (c *Cache) updateObj(domainMap map[string]CRMap, item *v1.AthenzDomain) {
+	c.lock.Lock()
+	domainName := item.ObjectMeta.Name
+	_, ok := domainMap[domainName]
+	if ok {
+		delete(domainMap, domainName)
+	}
+	roleToPrincipals := make(map[string][]*zms.RoleMember)
+	roleToAssertion := make(map[string][]SimpleAssertion)
+	crMap := CRMap{
+		RoleToPrincipals: roleToPrincipals,
+		RoleToAssertion:  roleToAssertion,
+	}
+	domainMap[domainName] = crMap
+	parseData(domainMap, domainName, item)
+	c.lock.Unlock()
+}
+
+func (c *Cache) deleteObj(domainMap map[string]CRMap, item *v1.AthenzDomain) {
+	c.lock.Lock()
+	domainName := item.ObjectMeta.Name
+	_, ok := domainMap[domainName]
+	if ok {
+		delete(domainMap, domainName)
+	}
+	c.lock.Unlock()
+}
+
+func (c *Cache) authorize(ctx context.Context, principal string, check AthenzAccessCheck) {
+	domainName := strings.Split(check.Resource, ":")
+	if len(domainName) != 2 {
+		fmt.Println("Error splitting domain name")
```

We need to return here, let's return an error.